sw/device/lib/base/print.c

```diff
@@ -0,0 +1,203 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/print.h"
+
+#include <stdarg.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+// Standard format specifiers.
+static const char kPercent = '%';
+static const char kCharacter = 'c';
+static const char kString = 's';
+static const char kSignedDec1 = 'd';
+static const char kSignedDec2 = 'i';
+static const char kUnsignedOct = 'o';
+static const char kUnsignedHexLow = 'x';
+static const char kUnsignedHexHigh = 'X';
+static const char kUnsignedDec = 'u';
+static const char kCharsSoFar = 'n';
+static const char kPointer = 'p';
+
+// Verilog-style format specifiers.
+static const char kSvBinary = 'b';
+static const char kSvHexLow = 'h';
+static const char kSvHexHigh = 'H';
+
+static const char kSizedStr = 'z';
+
+static const char kDigitsLow[16] = "0123456789abcdef";
+static const char kDigitsHigh[16] = "0123456789ABCDEF";
+
+static const char kErrorNul[17] = "%<unexpected nul>";
+static const char kUnknownSpec[15] = "%<unknown spec>";
+
+static void blackhole(const char *buf, size_t len) {}
+static buffer_sink_t base_stdout = &blackhole;
+
+void base_set_stdout(buffer_sink_t out) {
+  base_stdout = out;
+}
+
+size_t base_printf(const char *format, ...) {
+  va_list args;
+  va_start(args, format);
+  return base_core_printf(base_stdout, format, args);
+}
+
+size_t base_fprintf(buffer_sink_t out, const char *format, ...) {
+  va_list args;
+  va_start(args, format);
+  return base_core_printf(out, format, args);
+}
+
+static size_t write_digits(buffer_sink_t out, uint32_t value, uint32_t width,
+                           uint32_t base, const char *glyphs) {
+  char buffer[32];
+  size_t len = 0;
+  while (value > 0) {
+    uint32_t digit = value % base;
+    value /= base;
+    buffer[31 - len] = glyphs[digit];
+    ++len;
+  }
+  width = width == 0 ? 1 : width;
+  width = width > 32 ? 32 : width;
+  while (len < width) {
+    buffer[31 - len] = '0';
+    ++len;
+  }
+  out(buffer + (32 - len), len);
+  return len;
+}
+
+size_t base_core_printf(buffer_sink_t out, const char *format, va_list args) {
```

Comment:
If I have `out` set to `blackhole` (or a similar function that discards input rather than writing it anywhere), and I use `%n` in the format string, then the variable will be updated with the number of bytes "written", even though there is no output. 
- Is this surprising? (Should this be documented?)
- Is this a problem? (Should this be changed?) 

We cannot in general tell if `out` discards its outputs (without changing its signature to return the number of bytes it has written), which makes this harder.

Reply:
I was imagining `blackhole` to be like `/dev/null`, to which writes always succeed.

I think I'll change its name to that to make it more clear.

Reply:
Ok, then let's also document this in the top of `base_printf` (which removes some of the surprise)

Reply:
I added docs, I just was busy fixing stuff and hadn't pushed yet.